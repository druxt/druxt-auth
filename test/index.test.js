import DruxtAuthModule from '../src'

let mock

jest.mock('axios', () => ({
  post: jest.fn(() => ({
    data: true
  }))
}))

jest.mock('body-parser', () => ({
  json: () => jest.fn((req, res, fn) => {
    fn(req, res)
  })
}))

describe('DruxtAuth Nuxt module', () => {
  beforeEach(() => {
    mock = {
      addModule: jest.fn(),
      addTemplate: jest.fn(),
      extendRoutes: jest.fn(fn => {
        fn([], jest.fn())
      }),
      options: {
        druxt: {
          baseUrl: 'https://demo-api.druxtjs.org'
        },
        serverMiddleware: [],
      },
      DruxtAuthModule,
    }
  })

  test('Defaults', async () => {
    try {
      DruxtAuthModule.call(mock, {})
    } catch(err) {
      expect(err.message).toBe('DruxtAuth requires a clientId to be provided.')
    }

    // Call Druxt module with module options.
    DruxtAuthModule.call(mock, {
      clientId: 'mock-client-id'
    })

    // Expect the @nuxtjs/auth-next module to be correctly configured.
    expect(mock.options.auth).toMatchSnapshot()

    // Password grant middleware.
    expect(mock.options.serverMiddleware[0].path).toBe('/_auth/drupal-password/token')

    // Expect the middleware to ignore anything that isn't a POST request.
    let req = {}
    const res = {
      end: jest.fn()
    }
    const next = jest.fn()
    await mock.options.serverMiddleware[0].handler(req, res, next)
    expect(next).toHaveBeenCalled()

    // Expect username and password to be required.
    req = {
      method: 'POST',
      body: {
        grant_type: 'password'
      }
    }
    await mock.options.serverMiddleware[0].handler(req, res, next)
    expect(next).toHaveBeenCalledWith(new Error('Invalid username or password'))

    req = {
      method: 'POST',
      body: {
        grant_type: 'password',
        username: 'admin',
        password: 'password'
      }
    }
    await mock.options.serverMiddleware[0].handler(req, res, next)
    expect(res.end).toBeCalledWith('true')
  })
})
