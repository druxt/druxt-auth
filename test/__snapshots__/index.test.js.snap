// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DruxtAuth Nuxt module API Proxy - Array 1`] = `
Object {
  "redirect": Object {
    "callback": "/callback",
    "logout": "/",
  },
  "strategies": Object {
    "drupal-authorization_code": Object {
      "clientId": "mock-client-id",
      "codeChallengeMethod": "S256",
      "endpoints": Object {
        "authorization": "https://demo-api.druxtjs.org/oauth/authorize",
        "token": "https://demo-api.druxtjs.org/oauth/token",
        "userInfo": "/oauth/userinfo",
      },
      "grantType": "authorization_code",
      "responseType": "code",
      "scheme": "oauth2",
      "scope": undefined,
    },
    "drupal-password": Object {
      "endpoints": Object {
        "login": Object {
          "baseURL": "",
          "url": "/_auth/drupal-password/token",
        },
        "logout": false,
        "refresh": Object {
          "baseURL": "",
          "url": "/_auth/drupal-password/token",
        },
        "token": "https://demo-api.druxtjs.org/oauth/token",
        "user": Object {
          "method": "post",
          "url": "/oauth/userinfo",
        },
      },
      "grantType": "password",
      "refreshToken": Object {
        "data": "refresh_token",
        "maxAge": 2592000,
        "property": "refresh_token",
      },
      "scheme": "refresh",
      "token": Object {
        "maxAge": 31536000,
        "name": "Authorization",
        "property": "access_token",
        "type": "Bearer",
      },
      "user": Object {
        "property": false,
      },
    },
  },
}
`;

exports[`DruxtAuth Nuxt module API Proxy - Object 1`] = `
Object {
  "redirect": Object {
    "callback": "/callback",
    "logout": "/",
  },
  "strategies": Object {
    "drupal-authorization_code": Object {
      "clientId": "mock-client-id",
      "codeChallengeMethod": "S256",
      "endpoints": Object {
        "authorization": "https://demo-api.druxtjs.org/oauth/authorize",
        "token": "https://demo-api.druxtjs.org/oauth/token",
        "userInfo": "/oauth/userinfo",
      },
      "grantType": "authorization_code",
      "responseType": "code",
      "scheme": "oauth2",
      "scope": undefined,
    },
    "drupal-password": Object {
      "endpoints": Object {
        "login": Object {
          "baseURL": "",
          "url": "/_auth/drupal-password/token",
        },
        "logout": false,
        "refresh": Object {
          "baseURL": "",
          "url": "/_auth/drupal-password/token",
        },
        "token": "https://demo-api.druxtjs.org/oauth/token",
        "user": Object {
          "method": "post",
          "url": "/oauth/userinfo",
        },
      },
      "grantType": "password",
      "refreshToken": Object {
        "data": "refresh_token",
        "maxAge": 2592000,
        "property": "refresh_token",
      },
      "scheme": "refresh",
      "token": Object {
        "maxAge": 31536000,
        "name": "Authorization",
        "property": "access_token",
        "type": "Bearer",
      },
      "user": Object {
        "property": false,
      },
    },
  },
}
`;

exports[`DruxtAuth Nuxt module API Proxy - default 1`] = `
Object {
  "redirect": Object {
    "callback": "/callback",
    "logout": "/",
  },
  "strategies": Object {
    "drupal-authorization_code": Object {
      "clientId": "mock-client-id",
      "codeChallengeMethod": "S256",
      "endpoints": Object {
        "authorization": "https://demo-api.druxtjs.org/oauth/authorize",
        "token": "https://demo-api.druxtjs.org/oauth/token",
        "userInfo": "/oauth/userinfo",
      },
      "grantType": "authorization_code",
      "responseType": "code",
      "scheme": "oauth2",
      "scope": undefined,
    },
    "drupal-password": Object {
      "endpoints": Object {
        "login": Object {
          "baseURL": "",
          "url": "/_auth/drupal-password/token",
        },
        "logout": false,
        "refresh": Object {
          "baseURL": "",
          "url": "/_auth/drupal-password/token",
        },
        "token": "https://demo-api.druxtjs.org/oauth/token",
        "user": Object {
          "method": "post",
          "url": "/oauth/userinfo",
        },
      },
      "grantType": "password",
      "refreshToken": Object {
        "data": "refresh_token",
        "maxAge": 2592000,
        "property": "refresh_token",
      },
      "scheme": "refresh",
      "token": Object {
        "maxAge": 31536000,
        "name": "Authorization",
        "property": "access_token",
        "type": "Bearer",
      },
      "user": Object {
        "property": false,
      },
    },
  },
}
`;

exports[`DruxtAuth Nuxt module Defaults 1`] = `
Object {
  "redirect": Object {
    "callback": "/callback",
    "logout": "/",
  },
  "strategies": Object {
    "drupal-authorization_code": Object {
      "clientId": "mock-client-id",
      "codeChallengeMethod": "S256",
      "endpoints": Object {
        "authorization": "https://demo-api.druxtjs.org/oauth/authorize",
        "token": "https://demo-api.druxtjs.org/oauth/token",
        "userInfo": "https://demo-api.druxtjs.org/oauth/userinfo",
      },
      "grantType": "authorization_code",
      "responseType": "code",
      "scheme": "oauth2",
      "scope": undefined,
    },
    "drupal-password": Object {
      "endpoints": Object {
        "login": Object {
          "baseURL": "",
          "url": "/_auth/drupal-password/token",
        },
        "logout": false,
        "refresh": Object {
          "baseURL": "",
          "url": "/_auth/drupal-password/token",
        },
        "token": "https://demo-api.druxtjs.org/oauth/token",
        "user": Object {
          "method": "post",
          "url": "https://demo-api.druxtjs.org/oauth/userinfo",
        },
      },
      "grantType": "password",
      "refreshToken": Object {
        "data": "refresh_token",
        "maxAge": 2592000,
        "property": "refresh_token",
      },
      "scheme": "refresh",
      "token": Object {
        "maxAge": 31536000,
        "name": "Authorization",
        "property": "access_token",
        "type": "Bearer",
      },
      "user": Object {
        "property": false,
      },
    },
  },
}
`;
